doc/mkdocs/docs/api/basic_json/erase.md

```diff
@@ -10,7 +10,9 @@ iterator erase(iterator first, iterator last);
 const_iterator erase(const_iterator first, const_iterator last);
 
 // (3)
-size_type erase(const typename object_t::key_type& key);
+template<typename KeyT>
+size_type erase(KeyT && key);
+size_type erase(const std::string_view& key); // since C++17
```

Comment:
I believe that the template function is sufficient to cover the string_view case as well.

Reply:
If I remove the special overload and change the template to

```cpp
    template < class KeyT, typename std::enable_if <
#if defined(JSON_HAS_CPP_17)
                   std::is_same<typename std::decay<KeyT>::type, std::string_view>::value ||
#endif
                   std::is_convertible<KeyT, typename object_t::key_type>::value, int >::type    = 0 >
    size_type erase(KeyT && key)
```

then I get the following error:

```
In file included from ../test/src/unit-element_access2.cpp:32:
../include/nlohmann/json.hpp:4478:36: error: no matching member function for call to 'erase'
            return m_value.object->erase(std::forward<KeyT>(key));
                   ~~~~~~~~~~~~~~~~^~~~~
../test/src/unit-element_access2.cpp:988:25: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::erase<std::__1::basic_string_view<char, std::__1::char_traits<char> >, 0>' requested here
                CHECK(j.erase(std::string_view("integer")) == 1);
                        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:1299:14: note: candidate function not viable: no known conversion from 'std::__1::basic_string_view<char, std::__1::char_traits<char> >' to 'std::__1::map<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >, std::__1::less<void>, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > > > >::const_iterator' (aka '__map_const_iterator<__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >, void *> *, long> >') for 1st argument
    iterator erase(const_iterator __p) {return __tree_.erase(__p.__i_);}
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:1301:14: note: candidate function not viable: no known conversion from 'std::__1::basic_string_view<char, std::__1::char_traits<char> >' to 'std::__1::map<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >, std::__1::less<void>, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > > > >::iterator' (aka '__map_iterator<__tree_iterator<std::__1::__value_type<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >, void *> *, long> >') for 1st argument
    iterator erase(iterator __p)       {return __tree_.erase(__p.__i_);}
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:1303:15: note: candidate function not viable: no known conversion from 'std::__1::basic_string_view<char, std::__1::char_traits<char> >' to 'const std::__1::map<std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >, std::__1::less<void>, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > > > >::key_type' (aka 'const std::__1::basic_string<char>') for 1st argument
    size_type erase(const key_type& __k)
              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:1306:15: note: candidate function not viable: requires 2 arguments, but 1 was provided
    iterator  erase(const_iterator __f, const_iterator __l)
              ^
1 error generated.
```

Reply:
Ok. I'll have to take a look at this later. A bit busy today.

Reply:
My guess on this one is that the `std::map::erase()` function isn't templated. It needs either an iterator, or a key_type.

So you'd want to use `std::map::find()` of the provided key first, then call erase.

Reply:
Yes, that erase isn't already templated was discussed in the issue. Only .find and .emplace are.